src/java/com/zimbra/cs/nginx/NginxLookupExtension.java

```diff
@@ -880,13 +880,23 @@ private DomainExternalRouteInfo getDomainExternalRouteInfo(ILdapContext zlc, Con
 
         private Server lookupUpstreamImapServer(NginxLookupRequest req) throws ServiceException {
             ImapLoadBalancingMechanism LBMech = ImapLoadBalancingMechanism.newInstance();
-            Server server = prov.getLocalServer();
-            String[] imapServerAddrs;
-            if (server != null) {
-                imapServerAddrs = server.getReverseProxyUpstreamImapServers();
-            } else {
-                imapServerAddrs = prov.getConfig().getReverseProxyUpstreamImapServers();
+            String[] imapServerAddrs = null;
+            Account acct = prov.get(AccountBy.name, req.user);
+            if (acct != null) {
+                imapServerAddrs = acct.getServer().getReverseProxyUpstreamImapServers();
+            }
+
+            if (imapServerAddrs == null || imapServerAddrs.length == 0){
```

Comment:
the desired outcome is that admins can put IMAP-heavy users onto a subset of mailbox servers and configure a set of IMAPd servers to handle IMAP traffic for these accounts by specifying zimbraReverseProxyUpstreamImapServers attribute in the LDAP config for those mailbox servers. For all other accounts, IMAP traffic will be handled, according to default behavior, by mailbox servers. Global configuration can be used as a simplified configuration approach for smaller deployments or  for more homogenous where all accounts are heavy IMAP users. In this logic, the value of zimbraReverseProxyImapServers assigned to the server handling NGINX Lookup call is not relevant.